### PR TITLE
[MRG] Find end pts of section and a numpy version of h.distance()

### PR DESCRIPTION
**Pull request by raj1701**

### CODE REVIEW

File: hnn_core/find_coords.py
Comment (raj1701) [inline hnn_core/find_coords.py:11]: The `end_pts` printed here are different from those defined in `cells_default` for cell `L2Pyr`. 
Screenshot from 2023-07-03 01 42 33

Do I simply need to replace these `end_pts` in `cells_default` or should I build the cell and find them using `cell._nrn_sections`

Comment (jasmainak) [reply hnn_core/find_coords.py:11]: You need to build the cell and get the coords from that

Comment (raj1701) [reply hnn_core/find_coords.py:11]: When any new cell is created using `Cell` class, a private function `__update_end_pts()` is called at the end of `__init__`. This function creates neuron sections by calling `_create_sections()` then updates the `end_pts` of the `section`. The neuron objects are then deleted using `del self._nrn_sections[name]`.

I have a few questions regarding this flow -
- Neuron objects are created and destroyed in this flow which according to the discussion in last meeting should be avoided outside of simulation.
- If this functionality is already there and the `end_pts` are automatically updated, is manually updating the `end_pts` required ?
- Should this function `_update_end_pts()` be removed?

Comment (raj1701) [reply hnn_core/find_coords.py:11]: Also if `neuron objects` can be cleared using `del`, can't we follow a similar scheme while saving the network? After calling `build` on the copy of a cell, can we just iteratively delete the `_nrn_sections` and `_nrn_synapses` of the built cell??

Comment (jasmainak) [reply hnn_core/find_coords.py:11]: The problem is that Neuron objects are wrappers around C objects ... when `del` is called, it makes a request to the garbage collector to clear the object from memory. It's a request, not an order. If a pointer is not cleared from memory because of how it is coded in C, it will lead to a segmentation fault and/or slow down the following simulation. The latter will be very hard to debug ... we do have a scheme to clear neuron objects in `NetworkBuilder` but it was contributed by someone really in the weeds who understands Neuron. 

But you have a good point about `_update_end_pts`  ... I think @ntolley implemented that. Not sure how well it's been tested if the `del` in that function does not lead to a performance regression. Any thoughts Nick?

**Comment by raj1701** [comment]
As discussed in the meeting, I stored a base network and dipole and checked equality of them after removing `_update_end_pts()` and `h.define_shape()` .  Here are some observations -
- It seems that both `_update_end_pts()` and `h.define_shape()` are required in order to obtain the final `end_pts`. Removal of either one of them lead to a mismatch
- The reason is `_update_end_pts()` is dependent on `h.define_shape()`. Also the `end_pts` are only updated once when the cell is initialised. This function is only used by `modify_section()` which itself is not used anywhere.

**Comment by raj1701** [comment]
I think if all `end_pts` are updated in `cells_default`, building any neuron objects before simulation may be avoided. I will try this now.

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:783]: clearly explain the data structure

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:842]: method name appropriate?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:369]: simplify ...

**Comment by jasmainak** [comment]
@raj1701 let me know when I should look !

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:207]: you need to document it

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:341]:
```suggestion
```

comment is pretty redundant with the line below

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:301]:
```suggestion
        where sec_name is the name of the section and sec_end_pt is the 0 end
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:300]:
```suggestion
        Root is the 0 end of 'soma'. Nodes are a tuple (sec_name, node_pos)
```

node_pos = node position

more correct / easier to understand?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:893]: just use `np`

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:815]: if we got rid of `self.topology` would we need this function? how does the `self.cell_tree` look for a pyramidal cell? or should we still keep `self.topology`? They seem to be redundant ...

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:206]: For this function as well as any relevant ones below, can you add a comment with a URL to the relevant C++ code in NEURON? Think it will be useful to have the origin of this documented

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:366]:
```suggestion
    def distance_section(self, target_sec_name, base_node):
```

to be consistent with the tree representation? seems to simplify lot of stuff below if you stick to well-defined data structure

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:367]:
```suggestion
        if base_node not in self.cell_tree:
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:371]:
```suggestion
        curr_sec_children = self.cell_tree[base_node]
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:298]: what are the keys and values of the dictionary?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:387]:
```suggestion
        for node in self.cell_tree[base_node]:
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:393]:
```suggestion
                dist_1 = self.distance_section(target_sec_name, node)
```

I don't fully follow the logic of these if/else statements

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:910]:
```suggestion
        test_node = (sec_name, flag_start)
        if test_node in self.cell_tree:
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:911]:
```suggestion
            for child_node in self.cell_tree[test_node]:
```

**Review by jasmainak — CHANGES_REQUESTED** [review]
I spent some time reading the code ... everything isn't fully clear to me yet but it's getting there! Let's iterate a bit more so we can clean it up nicely. I'm passing the ball back to you @raj1701 . Let me know as soon as you do your pass (in a day or two?) and I'll read it one more time

**Comment by rythorpe** [comment]
This looks like a recreation of the NEURON functionality, where the cell sections are first defined with endpoints, attached to each other into a "tree", and then morphed according a specified length parameter for each section. This is exactly what we're trying to avoid, which can be accomplished by specifying the final endpoints to begin with.

**Comment by jasmainak** [comment]
@rythorpe this was done so as to avoid calling Neuron functions when saving the network ... right now, it's all implemented in pure python, so the ionic concentrations are known without using Neuron. I remember you were concerned about dangling neuron objects. This takes care of that issue. Furthermore, we better understand how the neuron distance function works :)

**Comment by rythorpe** [comment]
Yes, I thought the whole goals was just to update the hardcoded endpoints because it'd be the most lightweight solution. I don't understand why we needed to re-implement everything in Python.

**Comment by jasmainak** [comment]
hardcoding wouldn't have achieved what you wanted because of:

https://github.com/jonescompneurolab/hnn-core/blob/c81e5b87fbcc13bd4c6e5a403e838ca63ea16760/hnn_core/cell.py#L682-L697

see it already calls neuron functions. You'd have deprecate this ...

File: hnn_core/tests/test_cell.py
Comment (ntolley) [inline hnn_core/tests/test_cell.py:120]: Why are we dividing by two after calling update endpoints? wouldn't this lead to a mismatch?

Comment (ntolley) [reply hnn_core/tests/test_cell.py:120]: Definitely think this is the right format for the test, and the explicit checks below are good too, but not sure I totally follow all the logic

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:206]: Sure will add them

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:300]: Should I rename all instances of `sec_end_pt` as `node_pos`?

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:887]: Think of making the indexing more explicit, even with if/else statements, rather than boolean logic

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:393]: The algorithm searches for the destination nodes (end 0 or 1 of the target section) in all available paths. If the target section is in the subtree of the current node dist_1 (now named it `dist_temp`) will contain distance between the current node to the centre of the target section

File: hnn_core/tests/test_cell.py
Comment (raj1701) [reply hnn_core/tests/test_cell.py:120]: We divide all section lengths by 2 to get back the original end points because earlier we had doubled them and updated end_pts in line 117.

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:887]: Added this

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:890]: Admittedly uglier but thank you for humoring me :1st_place_medal:

File: hnn_core/cells_default.py
Comment (jasmainak) [inline hnn_core/cells_default.py:194]: for sake of clarity, do you mind separating the connections where the section name is same with those that connect different sections? Something like:

```py
    cell_tree = {
        ('apical_1', 0): [('apical_1', 1)],
        ('basal_1', 0): [('basal_1', 1)],
        ('apical_2', 0): [('apical_2', 1)],
        ('apical_tuft', 0): [('apical_tuft', 1)],
        ('apical_oblique', 0): [('apical_oblique', 1)],
        # different sections connected
        ('soma', 0): [('soma', 1), ('basal_1', 0)],
        ('soma', 1): [('apical_trunk', 0)],
        ('apical_trunk', 0): [('apical_trunk', 1)],
        ('apical_trunk', 1): [('apical_1', 0), ('apical_oblique', 0)],
        ('apical_1', 1): [('apical_2', 0)],
        ('apical_2', 1): [('apical_tuft', 0)]
}
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:459]: can this logic be moved here: https://github.com/jonescompneurolab/hnn-core/blob/master/hnn_core/cells_default.py ?

So we don't use the `partial` in the first place and directly store the values when creating the cell.

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:450]: any chance we can use `np.linspace` ? That seems exactly for this purpose

File: hnn_core/cells_default.py
Comment (jasmainak) [inline hnn_core/cells_default.py:99]: same comment as https://github.com/jonescompneurolab/hnn-core/pull/661/files#r1300563151

File: hnn_core/tests/test_cell.py
Comment (jasmainak) [inline hnn_core/tests/test_cell.py:142]:
```suggestion
```

remove any commented lines to make ready for merge !

File: hnn_core/tests/test_cell.py
Comment (jasmainak) [inline hnn_core/tests/test_cell.py:144]:
```suggestion
    # Checking equality till 5 decimal places
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:823]: needs to be indented

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:870]:
```suggestion
        """Update all end pts according to the length of the sections.
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:427]: let's be brave and delete `_set_biophysics`

Comment (rythorpe) [reply hnn_core/cell.py:427]: Don't we still need this function for inserting nrn mechanisms and such?

Comment (jasmainak) [reply hnn_core/cell.py:427]: I think `_set_section_mechs` is doing that now. This method only exists for checking equality with the old stuff

File: hnn_core/cells_default.py
Comment (raj1701) [reply hnn_core/cells_default.py:194]: Sure no problem

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:450]: Very slight precision errors are there. But it is passing the check in `_set_biophysics()`

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:427]: I think it is still required as it is inserting mechanisms in the NEURON section objects.
`_set_section_mechs()` is just expanding the partial functions for now. If we include the `_set_biophysics()` logic here then NEURON objects might still be created before simulating the network (in case of network write)

**Comment by ntolley** [comment]
image
I'm a bit stumped as to why this is only failing on windows...

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:431]:
```suggestion
    def _compute_section_mechs(self):
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:461]:
```suggestion
```

**Comment by jasmainak** [comment]
@raj1701 please rebase this PR so that the CIs pass

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:450]:
```suggestion
            seg_centers = np.linspace(0, 1, section.nseg * 2 + 1)[1::2]
```

does this work?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:213]: you need to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) your code:

https://github.com/raj1701/hnn-core/blob/neuron_coordinates/hnn_core/cell.py#L504-L508

make a private function `_get_nseg` that you use in both parts of the code.

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:540]: is this for testing purposes? Do you need it?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:367]: make private since not documented?

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:346]: do you still need the method `_update_end_pts` ? We should consistently update everywhere and remove the old method if not needed

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:814]:
```suggestion
        """Redefines end_pts according to section lengths.

        Detects change in section lengths of the sections in the
```

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:869]:
```suggestion
        """Update all end pts according to the length of the sections.
 
        Can be used whenever length of any section is updated
```

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:268]:
```suggestion
        list of nodes (children nodes) connected to the parent node.
```

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:306]:
```suggestion
        list of nodes (children nodes) connected to the parent node.
```

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:777]: To be clear this function should be deleted right?

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:398]:
```suggestion
            dist = np.nanmin(dist, dist_temp)
```

File: hnn_core/cell.py
Comment (ntolley) [inline hnn_core/cell.py:374]:
```suggestion
            return np.nan
```

**Comment by jasmainak** [comment]
@raj1701 let us know when we can look again!

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:450]: Yes it does

File: hnn_core/cell.py
Comment (raj1701) [reply hnn_core/cell.py:398]: It throws a warning when all nan values are passed to the function. The results are correct. This warning is fine right?

Comment (raj1701) [reply hnn_core/cell.py:398]: For now I have added a check to resolve this issue.

**Review by jasmainak — APPROVED** [review]
Looks good to me! @rythorpe or @ntolley do you want to merge?

**Comment by rythorpe** [comment]
I won't be able to look this over until tomorrow afternoon. If we don't feel like waiting until then, feel free to merge without me @ntolley.

**Comment by jasmainak** [comment]
I went ahead and merged to keep the momentum! @raj1701 let's bring your other PR to merge state as soon as possible :)

**Pull request merged**
